File: articles/cell-signaling-communication-biological-coordination/python/signaling_decay.py
"""
Signaling decay and half-life estimation.

Run:
    python python/signaling_decay.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DECAY_PATH = ARTICLE_DIR / "data" / "signaling_decay.csv"


def main() -> None:
    """Fit a simple exponential signaling decay model."""

    df = pd.read_csv(DECAY_PATH)

    time_min = df["time_min"].to_numpy(dtype=float)
    signal = df["signal"].to_numpy(dtype=float)

    slope, intercept = np.polyfit(time_min, np.log(signal), 1)

    k_est = -slope
    m0_est = np.exp(intercept)
    half_life = np.log(2) / k_est

    df["predicted_signal"] = np.exp(intercept + slope * time_min)
    df["residual"] = df["signal"] - df["predicted_signal"]

    summary = pd.DataFrame(
        {
            "k_est": [k_est],
            "m0_est": [m0_est],
            "half_life_min": [half_life],
        }
    )

    print(summary.round(4).to_string(index=False))
    print(df.round(4).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/cell-signaling-communication-biological-coordination/r/signaling_decay.R
# Signaling decay and half-life estimation in R.

decay_path <- file.path("data", "signaling_decay.csv")

if (!file.exists(decay_path)) {
  decay_path <- file.path("..", "data", "signaling_decay.csv")
}

decay_df <- read.csv(decay_path)

fit <- lm(log(signal) ~ time_min, data = decay_df)

k_est <- -coef(fit)[["time_min"]]
m0_est <- exp(coef(fit)[["(Intercept)"]])
half_life <- log(2) / k_est

summary_df <- data.frame(
  k_est = k_est,
  m0_est = m0_est,
  half_life_min = half_life
)

decay_df$predicted_signal <- exp(predict(fit))
decay_df$residual <- decay_df$signal - decay_df$predicted_signal

print(round(summary_df, 4))
print(round(decay_df, 4))
